pallets/creditcoin/src/types/transfer.rs

```diff
@@ -0,0 +1,149 @@
+use super::*;
+use crate::ocw::errors::SchedulerError;
+use crate::ocw::tasks::OffchainVerification;
+use crate::types::concatenate;
+use crate::Config;
+use crate::TaskOutput;
+use pallet_offchain_task_scheduler::tasks::error::TaskError;
+use pallet_offchain_task_scheduler::tasks::TaskV2;
+use pallet_timestamp::Config as TimestampConfig;
+use sp_runtime::traits::Hash;
+use sp_runtime::traits::UniqueSaturatedFrom;
+use sp_runtime::traits::UniqueSaturatedInto;
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct Transfer<AccountId, BlockNum, Hash, Moment> {
+	pub blockchain: Blockchain,
+	pub kind: TransferKind,
+	pub from: AddressId<Hash>,
+	pub to: AddressId<Hash>,
+	pub deal_order_id: DealOrderId<BlockNum, Hash>,
+	pub amount: ExternalAmount,
+	pub tx_id: ExternalTxId,
+	pub block: BlockNum,
+	pub is_processed: bool,
+	pub account_id: AccountId,
+	pub timestamp: Option<Moment>,
+}
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct UnverifiedTransfer<AccountId, BlockNum, Hash, Moment> {
+	pub transfer: Transfer<AccountId, BlockNum, Hash, Moment>,
+	pub from_external: ExternalAddress,
+	pub to_external: ExternalAddress,
+	pub deadline: BlockNum,
+	pub currency_to_check: CurrencyOrLegacyTransferKind,
+}
+
+impl<AccountId, BlockNum, Hash, Moment> UnverifiedTransfer<AccountId, BlockNum, Hash, Moment>
+where
+	Moment: UniqueSaturatedInto<u64> + UniqueSaturatedFrom<u64>,
+	BlockNum: UniqueSaturatedInto<u64>,
+{
+	pub fn into_output<T: Config>(
+		self,
+		timestamp: Option<T::Moment>,
+	) -> Transfer<AccountId, BlockNum, Hash, Moment>
+	where
+		T: Config<AccountId = AccountId, BlockNumber = BlockNum, Hash = Hash, Moment = Moment>,
+	{
+		Transfer { timestamp, ..self.transfer }
+	}
+}
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct TransferId<Hash>(Hash);
+
+impl<H> TransferId<H> {
+	#[cfg(test)]
+	pub fn leaked_inner_hash<C: SystemConfig>(blockchain: &Blockchain, blockchain_tx_id: &[u8]) -> H
+	where
+		<C as SystemConfig>::Hashing: Hash<Output = H>,
+	{
+		Self::inner_hash::<C::Hashing>(blockchain, blockchain_tx_id)
+	}
+
+	fn inner_hash<Hasher>(blockchain: &Blockchain, blockchain_tx_id: &[u8]) -> H
+	where
+		Hasher: Hash<Output = H>,
+	{
+		let key = concatenate!(&*blockchain.as_bytes(), blockchain_tx_id);
+		<Hasher as Hash>::hash(&key)
+	}
+
+	pub fn new<C: SystemConfig>(blockchain: &Blockchain, blockchain_tx_id: &[u8]) -> TransferId<H>
+	where
+		<C as SystemConfig>::Hashing: Hash<Output = H>,
+	{
+		let hash = Self::inner_hash::<C::Hashing>(blockchain, blockchain_tx_id);
+		TransferId(hash)
+	}
+}
+
+impl<H> From<H> for TransferId<H> {
```

Comment:
Can we add a `#[cfg(test)]` to this impl? There isn't really a valid reason to go straight from a raw Hash to TransferId outside of tests

Reply:
I want to but, it is hard to give up this pattern. Any 'pending task' can use Task::to_id() -> Config::Hash. And after you are done scheduling the task, you can From<Config::Hash> to generate an appropriate ItemId to query its storage. It is not that uncommon as a pattern, it does give the caller more responsability though, kind of what you would do with downcasting or memory transmuting. Any Hash created from its pending task is a valid ItemId but the reverse is not true. I was very careful to make clean usage of it, and it is never moved across scopes. I prefer this instead of having many ways of creating ids. The scheduler pallet does not care about ItemId specifics, it only knows how to handle Config::Hash, it is up to the pallet user to go from Config::Hash to an ItemId only when it makes sense. 

If you can come up with a cleaner implementation without spilling logic I'll give in lol:
[register_transfer_internal](https://github.com/gluwa/creditcoin/pull/773/files/595a84159cad516a6ec0be4f344e449e924ac407#diff-5de4052f8765736b266190fe0c2651e96120142610883cd5ca416a812593ecdfR104-R108)
```rust
let inner_id = TaskV2::<T>::to_id(&pending);

Self::check_and_submit_transfer_as_task(&inner_id, pending)?;

Ok((inner_id.into(), transfer))
```